### PR TITLE
Replace *jwt.validateClaimsWithLeeway with custom validation func

### PR DESCRIPTION
**Pull request by sergiught**
### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### 🔧 Changes

We are replacing the *jwt.validateClaimsWithLeeway func with our own logic to fix the issue with multiple audiences within the 3rd party pkg as it seems we are not getting a new release. This was preferred right now to swapping the jwt library used under the hood as that would be a bigger undertaking that we'll defer for later, perhaps a v3. 

### 📚 References

- https://github.com/auth0/go-jwt-middleware/issues/148

### CODE REVIEW

File: examples/http-jwks-example/main_test.go [hunk 1]
```diff
@@ -73,12 +73,12 @@ func TestHandler(t *testing.T) {
...
 			}
 
-			token := buildJWTForTesting(t, jwk, testServer.URL, test.subject, []string{})
+			token := buildJWTForTesting(t, jwk, testServer.URL, test.subject, []string{"my-audience"})
 			req.Header.Set("Authorization", "Bearer "+token)
 
 			rr := httptest.NewRecorder()
 
-			mainHandler := setupHandler(testServer.URL, []string{})
+			mainHandler := setupHandler(testServer.URL, []string{"my-audience"})
```
Comment (sergiught) [inline examples/http-jwks-example/main_test.go:81]: This was incorrectly set to validate no audience.

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -26,5 +26,5 @@ jobs:
         uses: codecov/codecov-action@v3
         with:
           files: coverage.out
-          fail_ci_if_error: true
+          fail_ci_if_error: false
```
Comment (sergiught) [inline .github/workflows/test.yaml:29]: Quick toggle to not fail the CI build if we fail to upload codecov.

File: validator/validator.go [hunk 3]
```diff
@@ -99,62 +99,114 @@ func (v *Validator) ValidateToken(ctx context.Context, tokenString string) (inte
 		return nil, fmt.Errorf("could not parse the token: %w", err)
 	}
 
-	if string(v.signatureAlgorithm) != token.Headers[0].Algorithm {
-		return nil, fmt.Errorf(
-			"expected %q signing algorithm but token specified %q",
-			v.signatureAlgorithm,
-			token.Headers[0].Algorithm,
-		)
+	if err = validateSigningMethod(string(v.signatureAlgorithm), token.Headers[0].Algorithm); err != nil {
```
Comment (sergiught) [inline validator/validator.go:102]: Trying to refactor this func so that the steps read better.

File: validator/validator.go [hunk 4]
```diff
@@ -99,62 +99,114 @@ func (v *Validator) ValidateToken(ctx context.Context, tokenString string) (inte
...
-		validatedClaims.RegisteredClaims.NotBefore = registeredClaims.NotBefore.Time().Unix()
+	foundAudience := false
+	for _, value := range expectedClaims.Audience {
+		if actualClaims.Audience.Contains(value) {
+			foundAudience = true
+			break
+		}
+	}
+	if !foundAudience {
+		return jwt.ErrInvalidAudience
```
Comment (sergiught) [inline validator/validator.go:153]: This is mainly the fix for the multiple audiences described in #148

File: validator/validator.go [hunk 5]
```diff
@@ -99,62 +99,114 @@ func (v *Validator) ValidateToken(ctx context.Context, tokenString string) (inte
...
 		},
+		CustomClaims: customClaims,
 	}
 
-	if registeredClaims.Expiry != nil {
-		validatedClaims.RegisteredClaims.Expiry = registeredClaims.Expiry.Time().Unix()
+	return validatedClaims, nil
+}
+
+func validateClaimsWithLeeway(actualClaims jwt.Claims, expected jwt.Expected, leeway time.Duration) error {
```
Comment (sergiught) [inline validator/validator.go:137]: This replaces https://github.com/square/go-jose/blob/c9ac459e06bc9a5ab7efd5d3b3f421d6de27401f/jwt/validation.go#L76.

File: validator/validator.go (hunk 3 above)
Comment (Widcket) [reply validator/validator.go:102]: 👏🏼

File: validator/validator.go (hunk 5 above)
Comment (Widcket) [reply validator/validator.go:137]: Should this also include the issuer and jti validation?

**Review by Widcket — CHANGES_REQUESTED** [review]
Looks good, just left a comment regarding the sub and jti claims.

File: validator/validator.go (hunk 5 above)
Comment (sergiught) [reply validator/validator.go:137]: Although that original func does include them, we were never actually using them for the validation. We're only setting as expected the audience, issuer and the time based claims. 

Re: 
- https://github.com/auth0/go-jwt-middleware/blob/master/validator/validator.go#L82-L85

Comment (Widcket) [reply validator/validator.go:137]: Note that in other SDKs we do validate the sub claim: https://github.com/auth0/Auth0.swift/blob/master/Auth0/IDTokenValidator.swift#L56

Comment (Widcket) [reply validator/validator.go:137]: Maybe we can add that validation?

Comment (Widcket) [reply validator/validator.go:137]: (if applicable, of course).

Comment (sergiught) [reply validator/validator.go:137]: Unfortunately we can't introduce that right now without a breaking change. The `expectedClaims` are a private internal property inside the validator and we set it with the passed in issuer and audience when constructing the validator. We'll have to change this func https://github.com/auth0/go-jwt-middleware/blob/master/validator/validator.go#L59 to allow for the subject as well. I'll note this down for a potential upcoming v3 perhaps, wdyt?

**Review by Widcket — APPROVED** [review]

**Pull request merged**